src/coreComponents/dataRepository/DataContext.hpp

```diff
@@ -0,0 +1,205 @@
+
+#include "common/DataTypes.hpp"
+#include "common/Logger.hpp"
+#include "xmlWrapper.hpp"
+
+namespace geos
+{
+
+namespace dataRepository
+{
+
+class Group;
+class WrapperBase;
+
+/// This abstract class stores data that helps to retrieve an object:
+/// - from which position in a file (if applicable), see DataFileContext,
+/// - where it is located in the data hierarchy, see GroupContext and WrapperContext.
+/// Typically, the target object contain an unique_ptr< DataContext > instance of this class.
+class DataContext
+{
+public:
+
+  /**
+   * @brief Construct a new DataContext object.
+   * @param objectName the target object name
+   * @param isDataFileContext true if this Context is a DataFileContext (see isDataFileContext for more infos)
+   */
+  DataContext( string const & objectName, bool isDataFileContext );
+
+  /**
+   * @return A string that mention all the known informations to retrieve from where the target
+   * object comes from.
+   */
+  virtual string toString() const = 0;
+
+  string getObjectName() const
+  { return m_objectName; }
+
+  /**
+   * @brief In some cases, we need to know if a DataContext is from a file. It means that it
+   * is a more user-friendly information compared to other DataContext classes.
+   * @return true if the context is from a file.
+   */
+  bool isDataFileContext() const
+  { return m_isDataFileContext; }
+
+  /**
+   * @brief Insert toString() result in a stream.
+   */
+  friend std::ostream & operator<<( std::ostream & os, const DataContext & dt );
+
+protected:
+
+  /// see getObjectName()
+  string const m_objectName;
+
+  /// see isDataFileContext()
+  bool const m_isDataFileContext;
+
+};
+
+/// Helps to know where a Group is in the hierarchy.
+/// See DataContext class for more infos.
+class GroupContext : public DataContext
+{
+public:
+
+  /**
+   * @brief Construct a new GroupContext object
+   * @param group see getGroup()
+   */
+  GroupContext( Group & group );
+
+  /**
+   * @brief Get the target Group object (wrapper's parent in case of a WrapperContext).
+   */
+  Group & getGroup() const;
+
+  /**
+   * @copydoc DataContext::toString()
+   */
+  virtual string toString() const;
+
+protected:
+
+  /**
+   * @brief Construct a new GroupContext object
+   * @param group see getGroup()
+   * @param objectName Target object name.
+   */
+  GroupContext( Group & group, string const & objectName );
+
+  /// see getGroup()
+  Group & m_group;
+
+};
+
+/// Helps to know the source context of a Wrapper in the hierarchy, or in the source file, if possible.
+/// See DataContext class for more infos.
+class WrapperContext final : public GroupContext
+{
+public:
+
+  /**
+   * @brief Construct a new WrapperContext object
+   */
+  WrapperContext( WrapperBase & wrapper );
+
+  /**
+   * @copydoc DataContext::toString()
+   */
+  virtual string toString() const;
+
+};
+
+/// Helps to know from where a Group or a Wrapper has been declared in its source file (a xml typically).
```

Comment:
Stores information to retrieve where a Group or Wrapper has been declared in the input source file (e.g. XML)